plugins/outputs/sql/sql.go

```diff
@@ -234,43 +349,40 @@ func (p *SQL) Write(metrics []telegraf.Metric) error {
 		for column, value := range metric.Tags() {
 			columns = append(columns, column)
 			values = append(values, value)
+
+			err := p.addColumnIfDoesNotExist(tablename, column, p.Convert.Text)
+
+			if err != nil {
+				return err
+			}
 		}
 
 		for column, value := range metric.Fields() {
 			columns = append(columns, column)
 			values = append(values, value)
-		}
 
-		sql := p.generateInsert(tablename, columns)
+			err := p.addColumnIfDoesNotExist(tablename, column, p.deriveDatatype(value))
 
-		switch p.Driver {
-		case "clickhouse":
-			// ClickHouse needs to batch inserts with prepared statements
-			tx, err := p.db.Begin()
-			if err != nil {
-				return fmt.Errorf("begin failed: %w", err)
-			}
-			stmt, err := tx.Prepare(sql)
 			if err != nil {
-				return fmt.Errorf("prepare failed: %w", err)
+				return err
 			}
-			defer stmt.Close() //nolint:revive,gocritic // done on purpose, closing will be executed properly
+		}
 
-			_, err = stmt.Exec(values...)
-			if err != nil {
-				return fmt.Errorf("execution failed: %w", err)
-			}
-			err = tx.Commit()
-			if err != nil {
-				return fmt.Errorf("commit failed: %w", err)
-			}
-		default:
-			_, err = p.db.Exec(sql, values...)
-			if err != nil {
-				return fmt.Errorf("execution failed: %w", err)
-			}
+		// sort the fields to improve the "Prepared" statement throughput
+		sortedColumns, sortedParams := p.sortFields(columns, values)
+
+		sql := p.generateInsert(tablename, sortedColumns)
+		batchMetrics[sql] = append(batchMetrics[sql], sortedParams)
+	}
+
+	for sql, sqlParams := range batchMetrics {
+		err := p.sendBatchTx(sql, sqlParams)
+
+		if err != nil {
+			return err
```

Comment:
I think this should be stripped out to another PR as it is in no way related to adding the columns...

Reply:
Sure